### PR TITLE
Relay the Java networking system properties to sub-processes to support proxy servers

### PR DESCRIPTION
**Pull request by olivierdagenais**
# Summary

Without these changes, a program using this component would open a web browser that would not follow the same proxy server configuration as the program and thus would be unable to complete the OAuth 2.0 Authorization Code Flow when access to the authorization endpoint needed to be performed through a proxy server.  This is because oauth2-useragent uses sub-processes to launch web browsers and the network-related configuration of the parent process was not propagated to the child processes.

This pull request propagates the values of networking-related system properties to the sub-processes such that if the main program can reach resources through a proxy server, so will the web browsers used for OAuth 2.0 flows.

# Manual testing

1. Set up a proxy server on the LAN.
    * I installed [Polipo](http://www.pps.univ-paris-diderot.fr/~jch/software/polipo/) on `192.168.0.117`.  It's listening on all interfaces at port `8123`.
1. Temporarily update the [Git Credential Manager for Mac and Linux](https://github.com/Microsoft/Git-Credential-Manager-for-Mac-and-Linux)'s pom.xml to point to the SNAPSHOT version of oauth2-useragent containing these changes.
1. Configure a VSTS account with 2-factor authentication.
1. For each of Windows 10, Mac OS X 10.10.5 and Fedora 22:
    1. Copy the SNAPSHOT build of GCM4ML containing these oauth2-useragent changes and configure Git to point to that version, enabling debug mode by setting `-Ddebug=true`.
    1. Configure the networking to prevent being able to reach web servers without going through a proxy server.
        * I did this by temporarily configuring the DHCP reservation to clear the `003 Router` option and then renewing the lease.
        * Running  `wget` yields `Network is unreachable` or `No route to host`.
    1. Clear or invalidate the saved credentials (if any) for the VSTS account, to force authentication.
        * Windows & Linux: I did this by renaming `insecureStore.xml` to `insecureStore.xml.old`.
        * Mac: I did this by deleting the corresponding Keychain entry.
    1. Run a `git clone` against a Git repo in the VSTS account.
        * Git fails saying it can't reach the host.
    1. Configure Git to use our proxy server by running:

        ```
        git config --global http.proxy http://192.168.0.117:8123
        ```
    1. Try the `git clone` again.
        * The GCM4ML emits the following:

            ```
            BaseVsoAuthentication::detectAuthority
                detected visualstudio.com, checking AAD vs MSA
                failed detection
                authority is basic
            ```
        ...which means it's unable to reach VSTS and so Git falls back to prompting for credentials.
    1. Update the `credential.helper` configuration to include the following:

        ```
        -Dhttps.proxyHost=192.168.0.117 -Dhttps.proxyPort=8123
        ```
    1. Try the `git clone` again.
        * Notice that this time, the GCM4ML "AAD vs MSA" check succeeds with `server has reponded` and the web browser is popped up to complete the authentication. :white_check_mark:
    1. Complete the authentication and notice the `git clone` succeeding.
    1. Windows & Linux ([Oracle documents this setting as only being available "\[on\] recent Windows systems and on Gnome 2.x systems"](http://docs.oracle.com/javase/7/docs/api/java/net/doc-files/net-properties.html))
        1. Update the `credential.helper` configuration to replace the two `https.proxy*` properties with the following:

            ```
            -Djava.net.useSystemProxies=true
            ```
        1. Configure the proxy server for the operating system. 
        1. Clear or invalidate the saved credentials for the VSTS account, to force authentication.
        1. Run `git fetch` inside the local copy of the Git repository.
            * Notice that the GCM4ML "AAD vs MSA" check succeeds again, the web browser again is popped up and, after completing the authentication, Git indeed communicates with the remote. :white_check_mark:
        1. Clear the proxy server configuration for the operating system.
    1. Undo the Git `http.proxy` configuration.
    1. Undo the networking configuration, so as to restore direct access to web servers.

Mission accomplished!

### CODE REVIEW

File: oauth2-useragent-core/src/test/java/com/microsoft/alm/oauth2/useragent/UserAgentImplTest.java [hunk 1]
```diff
@@ -181,6 +183,28 @@ private static void assertLinesMatch(final String actual, final String... expect
...
+        final HashSet<String> propertyNames = new HashSet<String>(Arrays.asList("http.proxyHost", "http.proxyPort"));
+        final Properties properties = new Properties();
+        properties.setProperty("http.proxyHost", "192.0.2.42");
+        final List<String> commands = new ArrayList<String>();
+
+        UserAgentImpl.relayProperties(properties, propertyNames, commands);
+
+        Assert.assertEquals(1, commands.size());
+        Assert.assertEquals("-Dhttp.proxyHost=192.0.2.42", commands.get(0));
+    }
```
Comment (DavidStaheli) [inline oauth2-useragent-core/src/test/java/com/microsoft/alm/oauth2/useragent/UserAgentImplTest.java]: Perhaps it's worth adding a test for all supported properties to catch any changes to the supported set.

Comment (olivierdagenais) [inline oauth2-useragent-core/src/test/java/com/microsoft/alm/oauth2/useragent/UserAgentImplTest.java]: I'm not sure I understand the suggestion.  Do you mean we add a copy of the contents of `NETWORKING_PROPERTY_NAMES` in a unit test that asserts that `NETWORKING_PROPERTY_NAMES` is equal to the copy?

I think I'd rather craft an integration test that uses `org.littleshoot:littleproxy:1.1.0-beta1` like [the test I added in tfs-plugin](https://github.com/jenkinsci/tfs-plugin/pull/72/files#diff-229f91b0cb0af95f157d6bcc15e0673aR742) that sets up a proxy server, connects to a remote host and then asserts that the proxy server was used.  What do you think?

File: oauth2-useragent-core/src/main/java/com/microsoft/alm/oauth2/useragent/UserAgentImpl.java [hunk 2]
```diff
@@ -116,6 +144,15 @@ AuthorizationResponse encode(final String methodName, final String... parameters
         }
     }
 
+    static void relayProperties(final Properties properties, final Set<String> propertyNames, final List<String> destinationCommand) {
+        for (final String propertyName : propertyNames) {
+            final String propertyValue = properties.getProperty(propertyName);
+            if (propertyValue != null) {
+                destinationCommand.add("-D" + propertyName + "=" + propertyValue);
```
Comment (yacaovsnc) [inline oauth2-useragent-core/src/main/java/com/microsoft/alm/oauth2/useragent/UserAgentImpl.java]: I think this looks good for now.  I am still unclear about how to work with a proxy that requires authentication.  Does passing "username:password@host" as the proxyHost work?

From what I read, it involves using `Authenticator`.  We can probably worry about that when there is a request.

Comment (olivierdagenais) [inline oauth2-useragent-core/src/main/java/com/microsoft/alm/oauth2/useragent/UserAgentImpl.java]: > I am still unclear about how to work with a proxy that requires authentication.

You're correct: I did not test that scenario.  The [Networking Properties](http://docs.oracle.com/javase/7/docs/api/java/net/doc-files/net-properties.html) page seems to indicate there's no way to configure HTTP and HTTPS proxy servers for authentication using system properties.

It's possible the `java.net.useSystemProxies` property would enable this transparently such that the proxy credentials wouldn't have to be visible in the configuration file as well as in process command-line arguments.

I've logged issue #18 to track this.

File: oauth2-useragent-core/src/test/java/com/microsoft/alm/oauth2/useragent/UserAgentImplTest.java (hunk 1 above)
Comment (olivierdagenais) [inline oauth2-useragent-core/src/test/java/com/microsoft/alm/oauth2/useragent/UserAgentImplTest.java]: > craft an integration test that uses org.littleshoot:littleproxy:1.1.0-beta1 like the test I added in tfs-plugin that sets up a proxy server, connects to a remote host and then asserts that the proxy server was used

I'll be implementing this; please hold off on merging until I follow up.

Thanks!
\- Oli

**Comment by olivierdagenais** [comment]
I was able to add two positive tests (they assert that the proxy server was indeed used) and I couldn't get a negative test going (configures the proxy, but turns it off, which should cause requests to fail), maybe because the JavaFX browser decides to try connecting directly if the proxy server isn't responding.
